File: cpp_ss/sandbox/mass_spring.cpp

#include <iostream>
#include <math.h>
#include <vector>

#include "blocks.hpp"
#include "helper.hpp"
#include "solver.hpp"
#include "gp-ios.hpp"

using namespace blocks;

class SSModel : public Submodel
{
public:
    SSModel() : Submodel("")
    {
        Node x("x");
        Node xd("xd");
        Node xdd("xdd");

        enter();
        {
            new Integrator("xd", xdd, xd, 0.1);
            new Integrator("x", xd, x);
            new Gain("-k/m", -1.0/1.0, x, xdd);
        }
        exit();
    }
};

int main()
{
    auto model = SSModel();
    auto history = run(model,
        [](uint k, double& t) -> bool
        {
            return arange(k, t, 0, 5, 0.01);
        },
        nullptr, NamedSignals(), rk4);

    Gnuplot gp;
	gp << "set xrange [0:500]\n";
    gp << "set yrange [-0.15:0.15]\n";
	gp << "plot" << gp.file1d(history["x"]) << "with lines title 'x',"
		<< gp.file1d(history["xd"]) << "with lines title 'xd'\n";

    return 0;
}

File: misc/py_ss/mass_spring.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import matplotlib.pyplot as plt
import numpy as np

import blocks, helper
from blocks import N as N

class SSModel(blocks.Submodel):
    def __init__(self):
        super().__init__('')

        x = N('x')
        xd = N('xd')

        with self:
            blocks.Integrator('xd', iport='-xdd', oport=xd, x0=0.1)
            blocks.Integrator('x', iport=xd, oport=x)
            blocks.Gain('-k/m', k=-1.0/1.0, iport=x, oport='-xdd')

def main():
    model = SSModel()
    history = helper.run(model, T=np.arange(0, 5, 0.01))

    T = history['t']
    plt.figure()
    plt.subplot(2, 1, 1); plt.plot(T, history['x'])
    plt.ylabel('x')
    plt.subplot(2, 1, 2); plt.plot(T, history['xd'])
    plt.ylabel('xd')
    plt.show()

if __name__ == '__main__':
    main()
